src/test/java/com/google/api/gax/bundling/PushingBundlerTest.java

```diff
@@ -0,0 +1,328 @@
+            return 1;
+          }
+        });
+  }
+
+  private static TrackedFlowController trackedFlowController;
+
+  private static BundlingFlowController<SimpleBundle> getTrackedIntegerBundlingFlowController(
+      Integer elementCount, Integer byteCount, LimitExceededBehavior limitExceededBehaviour) {
+    trackedFlowController =
+        new TrackedFlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(elementCount)
+                .setMaxOutstandingRequestBytes(byteCount)
+                .setLimitExceededBehavior(limitExceededBehaviour)
+                .build());
+    return new BundlingFlowController<>(
+        trackedFlowController,
+        new ElementCounter<SimpleBundle>() {
+          @Override
+          public long count(SimpleBundle t) {
+            return t.getIntegers().size();
+          }
+        },
+        new ElementCounter<SimpleBundle>() {
+          @Override
+          public long count(SimpleBundle t) {
+            long counter = 0;
+            for (Integer i : t.integers) {
+              counter += i;
+            }
+            return counter;
+          }
+        });
+  }
+
+  @Rule public ExpectedException thrown = ExpectedException.none();
+
+  private static class SimpleBundle {
+
+    private final List<Integer> integers = new ArrayList<>();
+
+    private static SimpleBundle fromInteger(Integer integer) {
+      SimpleBundle bundle = new SimpleBundle();
+      bundle.integers.add(integer);
+      return bundle;
+    }
+
+    public void merge(SimpleBundle t) {
+      integers.addAll(t.integers);
+    }
+
+    private List<Integer> getIntegers() {
+      return integers;
+    }
+  }
+
+  private static class SimpleBundleMerger implements BundleMerger<SimpleBundle> {
+    @Override
+    public void merge(SimpleBundle bundle, SimpleBundle newBundle) {
+      bundle.merge(newBundle);
+    }
+  }
+
+  private static PushingBundler.Builder<SimpleBundle> createSimpleBundlerBuidler(
+      AccumulatingBundleReceiver<SimpleBundle> receiver) {
+    return PushingBundler.<SimpleBundle>newBuilder()
+        .setThresholds(BundlingThresholds.<SimpleBundle>of(100))
+        .setExecutor(EXECUTOR)
+        .setMaxDelay(Duration.millis(10000))
+        .setReceiver(receiver)
+        .setFlowController(PushingBundlerTest.<SimpleBundle>getDisabledBundlingFlowController())
+        .setBundleMerger(new SimpleBundleMerger());
+  }
+
+  @Test
+  public void testAdd() throws Exception {
+    AccumulatingBundleReceiver<SimpleBundle> receiver = new AccumulatingBundleReceiver<>();
+    PushingBundler<SimpleBundle> bundler = createSimpleBundlerBuidler(receiver).build();
+    bundler.add(SimpleBundle.fromInteger(14));
+    Truth.assertThat(bundler.isEmpty()).isFalse();
+    Truth.assertThat(receiver.getBundles().size()).isEqualTo(0);
+
+    bundler.trigger();
+    // Give time for the executor to push the bundle
+    Thread.sleep(10);
```

Comment:
Made `pushCurrentBundle` return a future, and blocked on it in this case.

Note that other tests still contain sleeps - I think it would be quite hard to restructure to remove all of them.